areader/reader.go

```diff
@@ -16,313 +16,418 @@ package areader
 
 import (
 	"archive/tar"
+	"bytes"
 	"compress/gzip"
-	"fmt"
 	"io"
 	"os"
 	"path/filepath"
+	"strconv"
 	"strings"
 
-	"github.com/mendersoftware/mender-artifact/metadata"
-	"github.com/mendersoftware/mender-artifact/parser"
+	"github.com/mendersoftware/mender-artifact/artifact"
+	"github.com/mendersoftware/mender-artifact/handlers"
 	"github.com/pkg/errors"
 )
 
-type Reader struct {
-	r io.Reader
-	*parser.ParseManager
-
-	info    *metadata.Info
-	tReader *tar.Reader
-	*headerReader
-}
+type SignatureVerifyFn func(message, sig []byte) error
+type DevicesCompatibleFn func([]string) error
 
-type headerReader struct {
-	hInfo *metadata.HeaderInfo
-
-	hGzipReader *gzip.Reader
-	hReader     *tar.Reader
-	nextUpdate  int
+type Reader struct {
+	CompatibleDevicesCallback DevicesCompatibleFn
+	VerifySignatureCallback   SignatureVerifyFn
+
+	signed     bool
+	hInfo      *artifact.HeaderInfo
+	info       *artifact.Info
+	r          io.Reader
+	handlers   map[string]artifact.Installer
+	installers map[int]artifact.Installer
 }
 
 func NewReader(r io.Reader) *Reader {
 	return &Reader{
-		r:            r,
-		ParseManager: parser.NewParseManager(),
-		headerReader: &headerReader{hInfo: new(metadata.HeaderInfo)},
+		r:          r,
+		handlers:   make(map[string]artifact.Installer, 1),
+		installers: make(map[int]artifact.Installer, 1),
 	}
 }
 
-func isCompatibleWithDevice(current string, compatible []string) bool {
-	for _, dev := range compatible {
-		if strings.Compare(current, dev) == 0 {
-			return true
-		}
-	}
-	return false
+func (ar *Reader) isSigned() bool {
+	return ar.signed
 }
 
-func (ar *Reader) read(device string) (parser.Workers, error) {
-	defer func() { ar.tReader = nil }()
+func (ar *Reader) readHeader(tReader io.Reader) ([]byte, error) {
 
-	var err error
-	ar.info, err = ar.ReadInfo()
+	var r io.Reader
+	if ar.isSigned() {
+		// If artifact is signed we need to calculate header checksum to be
+		// able to validate it later.
+		r = artifact.NewReaderChecksum(tReader)
```

Comment:
Good point. We don't need to calculate checksum for v1 artifact though. I've changed the logic so that we are always calculating and verifying checksum for v2 artifact.